### ISSUE TITLE
Add Keyboard to Lichess natively?

### ISSUE THREAD

**Issue by mr-yt12** [mr-yt12/lichess-gui-use-code-example-new#44]
https://github.com/ornicar/chessground

Research about the possibility to implement KB / multipremove directly into Lichess. However, chessground code is already quite complex, so it should perhaps be in a separate, easily manageable component. I guess just KB could be a good start, and multipremoves could be added later? Or vice versa? If the player has it enabled, it only activates when playing against another player who has it enabled. It should have its own option in the Game Behavior settings (or two, one for KB and one for multipremoves). It should not be possible to change during a game, the changes will only be applied to the next games (to avoid people disabling it before the game and enabling it during the game to prevent the opponent from using it).

**Comment by mr-yt12** [comment]
@artyompal

**Comment by mr-yt12** [comment]
@artyompal, @randomguyreborn pinging you in case you have any suggestions

**Comment by randomguyreborn** [comment]
@mr-yt12 Yeah its possible to add kb to lichess natively  but the reason why they didn't allow kb is **Extensions that assist with playing moves have denatured fast chess to the point that many players won't play it anymore. We believe that banning keyboard extensions will result in UltraBullet and HyperBullet becoming more popular** so not sure whether they accept our  suggestions

**Comment by randomguyreborn** [comment]
@mr-yt12 also the sentence **it only activates when playing against another player who has it enabled.** okay this  is fine but only applies to pc and laptop players so what does phone or tab players do against their opponents?

**Comment by randomguyreborn** [comment]
@mr-yt12 Yeah its possible to add kb to lichess natively but the reason why they didn't allow kb is Extensions that assist with playing moves have denatured fast chess to the point that many players won't play it anymore. We believe that banning keyboard extensions will result in UltraBullet and HyperBullet becoming more popular so not sure whether they accept our suggestions, also the sentence it only activates when playing against another player who has it enabled. okay this is fine but only applies to pc and laptop players so what does phone or tab players do against their opponents?

**Comment by randomguyreborn** [comment]
@mr-yt12 Yeah its possible to add kb to lichess natively but the reason why they didn't allow kb is **Extensions that assist with playing moves have denatured fast chess to the point that many players won't play it anymore. We believe that banning keyboard extensions will result in UltraBullet and HyperBullet becoming more popular** so not sure whether they accept our suggestions,  also the sentence **it only activates when playing against another player who has it enabled**. okay this is fine but only applies to pc and laptop players so what does phone or tab players do against their opponents?

**Comment by randomguyreborn** [comment]
@mr-yt12 Yeah its possible to add kb to lichess natively but they didn't allow kb so not sure whether they accept our suggestions,  also the sentence **it only activates when playing against another player who has it enabled**. okay this is fine but only applies to pc and laptop players so what does phone or tab players do against their opponents?

**Comment by mr-yt12** [comment]
@randomguyreborn I asked, and I think the pull request will be merged if "it doesn't make the current code more complex, which would be surprising". Phone or tablet players could use multipremoves. However, to use KB, they would need a physical keyboard connected to their device. So if they have "use KB + hover" option disabled, they just play how they do now. However, if they have it enabled, they won't be able to input moves the standard way (touch and drag) (since I think using KB should disable all the other input methods, because otherwise it would require too much code complexity to address all the bugs which can result from someone, for example, dragging a piece, then making a KB move while the piece is being dragged), so maybe some warning could be added when the person enables it, explaining how it works.

**Comment by randomguyreborn** [comment]
@mr-yt12 are you working on this?

**Comment by mr-yt12** [comment]
@randomguyreborn no

**Comment by randomguyreborn** [comment]
@mr-yt12 ****I sometime might work on this also should the shortcut keys like for berserk and resign also need to be added?
but players other than pc or laptop might not get used to this keys****

**Comment by randomguyreborn** [comment]
@mr-yt12 This change https://github.com/ornicar/lila/commit/121bfc76ce69e3994e4ab7a870bf5478da0ee7a4 destroys the web sockets and redirect to https://lichess.org/page/play-extensions so it need to be modified to work

```scala
      if ($('#KeyboardO,#show_btn,#shadowHostId').length) {

**Comment by randomguyreborn** [comment]
https://github.com/mr-yt12/lichess-gui-use-code-example-new/blob/8c972071e1ed1365fae61557369800acf999e0bb/Old%20files/style.css#L1

**Comment by mr-yt12** [comment]
@randomguyreborn https://pastebin.com/bSrv4pVN (password: the code inside webstore kb bckg.js)

**Comment by mr-yt12** [comment]
@randomguyreborn 

'aHR0cHM6Ly9wYXN0ZWJpbi5jb20vYlNydjRwVk4gKHBhc3N3b3JkOiB0aGUgY29kZSBpbnNpZGUgd2Vic3RvcmUga2IgYmNrZy5qcyk='

**Comment by mr-yt12** [comment]
@randomguyreborn  

'QW5FbGVwaGFudFRoYXRTdGVwcGVkT25BQ29pbiM0MTg4'

**Comment by randomguyreborn** [comment]
@mr-yt12 do you think the short cut keys like back to tournament, resign also need to be added or should remove from the code?

**Comment by mr-yt12** [comment]
@randomguyreborn i guess they could be omitted for simplicity for now, and added later, in case the keys would interfere with some other shortcuts. Also some people might accidentally press them. So it's better to research it more, look into existing shortcut issues, or leave it for later
